contracts/extensions/ERC4907A.sol

```diff
@@ -0,0 +1,133 @@
+     * The zero address indicates that there is no user or if the user is expired.
+     */
+    function userOf(uint256 tokenId) public view returns (address) {
+        uint256 packed = _packedUserInfo[tokenId];
+        assembly {
+            // Branchless `packed *= block.timestamp <= expires ? 1 : 0`.
+            packed := mul(
+                packed,
+                // `block.timestamp <= expires ? 1 : 0`.
+                lt(
+                    shl(_BITPOS_EXPIRES, timestamp()),
+                    packed
+                )
+            )
+        }
+        return address(uint160(packed));
+    }
+
+    /**
+     * @dev Returns the user's expires of `tokenId`.
+     */
+    function userExpires(uint256 tokenId) public view returns (uint256) {
+        return _packedUserInfo[tokenId] >> _BITPOS_EXPIRES;
+    }
+
+    /**
+     * @dev See {IERC165-supportsInterface}.
+     */
+    function supportsInterface(bytes4 interfaceId) public view virtual override(ERC721A, IERC721A) returns (bool) {
+        // The interface ID for ERC4907 is `0xad092b5c`.
+        // See: https://eips.ethereum.org/EIPS/eip-4907
+        return super.supportsInterface(interfaceId) || interfaceId == 0xad092b5c;
+    }
+
+    /**
+     * @dev Returns the user address for `tokenId`, ignoring the expiry status.
+     */
+    function _explicitUserOf(uint256 tokenId) internal view returns (address) {
+        return address(uint160(_packedUserInfo[tokenId]));
+    }
+
+    /**
+     * @dev Overrides the `_beforeTokenTransfers` hook to clear the user info upon transfer.
+     */
+    function _beforeTokenTransfers(
+        address from,
+        address to,
+        uint256 startTokenId,
+        uint256 quantity
+    ) internal virtual override {
+        super._beforeTokenTransfers(from, to, startTokenId, quantity);
+
+        bool mayNeedClearing;
+        assembly {
+            // Branchless `quantity == 1 && !(from == address(0) || from == to)`.
+            // Saves 60+ gas.
+            // The addresses are masked with `_BITMASK_ADDRESS` to
+            // clear any non-zero excess upper bits.
+            mayNeedClearing := and(
+                eq(quantity, 1),
+                // `!(from == address(0) || from == to)`.
+                iszero(
+                    or(
+                        // Whether it is a mint (i.e. `from == address(0)`).
+                        iszero(and(from, _BITMASK_ADDRESS)),
+                        // Whether the owner is unchanged (i.e. `from == to`).
+                        eq(and(from, _BITMASK_ADDRESS), and(to, _BITMASK_ADDRESS))
+                    )
+                )
+            )
+        }
+
+        if (mayNeedClearing) {
+            // If either `user` or `expires` are non-zero.
+            if (_packedUserInfo[startTokenId] != 0) {
+                delete _packedUserInfo[startTokenId];
+                emit UpdateUser(startTokenId, address(0), 0);
+            }
+        }
+    }
```

Comment:
Since this function is called upon mint, transfer, and burn, it is very important to optimize it heavily.